include/openenclave/internal/posix/lock.h

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef _OE_POSIX_LOCK_H
+#define _OE_POSIX_LOCK_H
+
+#include <openenclave/enclave.h>
+
+#include <openenclave/corelibc/assert.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/thread.h>
+
+OE_EXTERNC_BEGIN
+
+/*
+**==============================================================================
+**
+** lock.h:
+**
+**     This file defines 'conditional lock' functions. These functions perform
+**     their operation based on the value of the locked parameter (see function
+**     descriptions below). The locked parameter must reside on the calling
+**     thread's stack. Here's a typical schenario.
+**
+**         static oe_spinlock_t _lock;
+**
+**         int f(const char* name)
+**         {
+**             int ret = -1;
+**             bool locked = false;
+**
+**             if (!name)
+**                 goto done;
+**
+**             oe_conditional_lock(&_lock, locked);
+**
+**             // Modify the lock protected structures here!
+**
+**         done:
+**             oe_conditional_unlock(&_lock, locked);
+**             return ret;
+**         }
+**
+**     Another scenario involves passing the locked variable onto another
+**     function. For example:
+**
+**         static oe_spinlock_t _lock;
+**
+**         int f(bool* locked)
```

Comment:
>f [](start = 15, length = 1)

Typo: should be `g(bool* locked)`

This is also a somewhat contrived example which doesn't illustrate why nested locking might be necessary. (i.e. `g()` can be called from `f()` and some `f'()` that doesn't lock prior to invoking `g()`)

Given the implementation in lock.c, it's also not safe given usage not illustrated by this example: `g()` will unlock the lock held by `f()`, which will violate mutex of the code that comes after `g()` returns before `f()` explicitly releases the lock.

Reply:
Backed out. Thanks!